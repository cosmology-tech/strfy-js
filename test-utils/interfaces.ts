export interface Asset {
  deprecated?: boolean;
  description?: string;
  extended_description?: string;
  denom_units: DenomUnit[];
  type_asset?: "sdk.coin" | "cw20" | "erc20" | "ics20" | "snip20" | "snip25" | "bitcoin-like" | "evm-base" | "svm-base" | "substrate";
  address?: string;
  base: string;
  name: string;
  display: string;
  symbol: string;
  traces?: (IbcTransition | IbcCw20Transition | NonIbcTransition)[];
  ibc?: {
    source_channel: string;
    dst_channel: string;
    source_denom: string;
  };
  logo_URIs?: {
    png?: string;
    svg?: string;
  };
  images?: {
    image_sync?: Pointer;
    png?: string;
    svg?: string;
    theme?: {
      primary_color_hex?: string;
      circle?: boolean;
      dark_mode?: boolean;
    };
  }[];
  coingecko_id?: string;
  keywords?: string[];
  socials?: {
    website?: string;
    twitter?: string;
  };
}
export interface DenomUnit {
  denom: string;
  exponent: number;
  aliases?: string[];
}
export interface Pointer {
  chain_name: string;
  base_denom?: string;
}
export interface IbcTransition {
  type: "ibc";
  counterparty: {
    chain_name: string;
    base_denom: string;
    channel_id: string;
  };
  chain: {
    channel_id: string;
    path: string;
  };
}
export interface IbcCw20Transition {
  type: "ibc-cw20";
  counterparty: {
    chain_name: string;
    base_denom: string;
    port: string;
    channel_id: string;
  };
  chain: {
    port: string;
    channel_id: string;
    path: string;
  };
}
export interface NonIbcTransition {
  type: "bridge" | "liquid-stake" | "synthetic" | "wrapped" | "additional-mintage" | "test-mintage";
  counterparty: {
    chain_name: string;
    base_denom: string;
    contract?: string;
  };
  chain?: {
    contract: string;
  };
  provider: string;
}
export interface AssetList {
  $schema?: string;
  chain_name: string;
  assets: Asset[];
}