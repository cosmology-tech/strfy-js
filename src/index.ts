export * from './js';
export * from './json';