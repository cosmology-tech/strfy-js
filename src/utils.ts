import { minimatch } from 'minimatch';

export function camelCaseTransform(key: string): string {
  return key.replace(/[-_\s]+(.)?/g, (_, c) => c ? c.toUpperCase() : '');
}

export function isSimpleKey(key: string): boolean {
  return /^[a-zA-Z_$][a-zA-Z0-9_$]*$/.test(key);
}

export function dirname(path: string): string {
  return path.replace(/\/[^\/]*$/, ''); // Removes last segment after the last '/'
}

export function escapeStringForSingleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only single quotes
  str = str.replace(/'/g, "\\'");
  return str;
}

export function escapeStringForDoubleQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape only double quotes
  str = str.replace(/"/g, '\\"');
  return str;
}


export function escapeStringForBacktickQuotes(str: string): string {
  // Escape backslashes first
  str = str.replace(/\\/g, '\\\\');
  // Escape control characters
  str = str.replace(/\n/g, '\\n');
  str = str.replace(/\r/g, '\\r');
  str = str.replace(/\t/g, '\\t');
  // Escape backticks
  str = str.replace(/`/g, '\\`');
  return str;
}


const globPattern = /\*+([^+@!?\*\[\(]*)/;

interface ShouldIncludeOptions {
  include: string[];
  exclude: string[];
}

export function expandPaths(paths: string[]): string[] {
  const pathSet = new Set<string>();

  paths.forEach(path => {
    let currentPath = path;
    // Ensure the path is added to the set
    pathSet.add(currentPath);

    // Generate and add all parent directories to the set
    while (currentPath !== '') {
      const lastSlashIndex = currentPath.lastIndexOf('/');
      if (lastSlashIndex === -1) break; // No more slashes, stop processing

      currentPath = currentPath.substring(0, lastSlashIndex) || '';
      if (currentPath !== '')
        pathSet.add(currentPath);
    }
  });

  return sortPaths(Array.from(pathSet));
}

function sortPaths(paths) {
  // First sort alphabetically
  paths.sort((a, b) => a.localeCompare(b));

  // Then sort by depth (number of '/' in the path)
  paths.sort((a, b) => {
    const depthA = (a.match(/\//g) || []).length;
    const depthB = (b.match(/\//g) || []).length;
    return depthB - depthA;  // Sort from deepest to shallowest
  });

  return paths;
}


function hasAncestor(currentPath, includePaths) {
  return !!findNearestAncestor(currentPath, includePaths);
}

function findNearestAncestor(currentPath, includePaths) {
  let pathToCheck = currentPath;
  while (pathToCheck !== '/' && pathToCheck !== '') {
    if (includePaths.some(includePath => minimatch(pathToCheck + '/', includePath))) {
      if (pathToCheck !== currentPath)
        return pathToCheck; // Return the matching parent directory
    }
    pathToCheck = dirname(pathToCheck); // Move to the next parent directory
  }
  return null; // Return null if no matching parent is found
}

export const shouldInclude = (type: string, options: ShouldIncludeOptions): boolean => {
  // Determine if 'include' and 'exclude' are effectively set
  const includesEffectivelySet = options.include && options.include.length > 0;
  const excludesEffectivelySet = options.exclude && options.exclude.length > 0;

  // Function to check if any patterns in the array match the type
  const matchesPattern = (patterns: string[], type: string): boolean =>
    patterns.some(pattern => globPattern.test(pattern) ? minimatch(type, pattern) : type === pattern);

  const isIncluded = includesEffectivelySet ? matchesPattern(expandPaths(options.include), type) : true;
  const isIncludedByAncestor = !!findNearestAncestor(type, options.include);
  const isExcluded = excludesEffectivelySet ? matchesPattern(options.exclude, type) : false;

  const toInclude = isIncluded;

  // Apply the logic based on whether includes or excludes are effectively set
  if (includesEffectivelySet && excludesEffectivelySet) {
    if (isIncludedByAncestor && !isExcluded) {
      return true;
    }
    return toInclude && !isExcluded;
  } else if (includesEffectivelySet) {
    if (isIncludedByAncestor) {
      return true;
    }
    return toInclude;
  } else if (excludesEffectivelySet) {
    return !isExcluded;
  }

  // Default behavior if neither is effectively set
  return true;
}