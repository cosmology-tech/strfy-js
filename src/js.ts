import { camelCaseTransform, dirname, escapeStringForBacktickQuotes, escapeStringForDoubleQuotes, escapeStringForSingleQuotes, isSimpleKey, shouldInclude } from "./utils";

export interface JSStringifySetterOptions<ObjType, RootType> {
  currentKey: string;
  currentPath: string;
  value: any;
  obj: ObjType;
  root: RootType;
  defaultValuesMap: { [path: string]: any };
};

export interface JSStringifyPropertyReplacerOptions<ObjType, RootType> {
  currentKey: string;
  currentPath: string;
  propertyRenameMap: { [pattern: string]: string };
  obj: ObjType;
  root: RootType
  value?: any;
}

export type JSStringifySetter = (options: JSStringifySetterOptions<any, any>) => any;
export type JSStringifyReplacer = (options: JSStringifyPropertyReplacerOptions<any, any>) => any;
export interface JSStringifyOptions {
  space?: number;
  propertyReplacer?: (options: JSStringifyPropertyReplacerOptions<any, any>) => string
  valueReplacer?: Record<string, (options: JSStringifyPropertyReplacerOptions<any, any>) => any>;
  quotes?: 'single' | 'double' | 'backtick';
  inlineArrayLimit?: number;
  camelCase?: boolean;
  camelCaseFn?: (str: string) => string;
  include?: string[];
  exclude?: string[];
  defaultValuesMap?: { [path: string]: any };
  propertyRenameMap?: { [path: string]: any };
  defaultValuesSetter?: { [path: string]: JSStringifySetter } | JSStringifySetter;
  json?: boolean;
}


export function chooseQuotes(str: string, preferred: 'single' | 'double' | 'backtick'): string {
  switch (preferred) {
    case 'single':
      return `'${escapeStringForSingleQuotes(str)}'`;
    case 'double':
      return `"${escapeStringForDoubleQuotes(str)}"`;
    case 'backtick':
      return `\`${escapeStringForBacktickQuotes(str)}\``;
    default:
      throw new Error("Invalid quote type specified.");
  }
}

function matchesPattern(path: string, pattern: string) {
  if (pattern === '*') return true;
  if (path === pattern) return true;
  const pathSegments = path.split('/');
  const patternSegments = pattern.split('/');
  if (pathSegments.length !== patternSegments.length) {
    return false;
  }
  for (let i = 0; i < pathSegments.length; i++) {
    if (patternSegments[i] !== '*' && patternSegments[i] !== pathSegments[i]) {
      return false;
    }
  }
  return true;
}

function defaultPropertyReplacer(
  currentKey: string,
  currentPath: string,
  propertyRenameMap: { [pattern: string]: string }
): string {
  let finalKey = currentKey;
  Object.keys(propertyRenameMap).forEach(pattern => {
    if (matchesPattern(currentPath, pattern)) {
      finalKey = propertyRenameMap[pattern];
    }
  });
  return finalKey;
}

export function jsStringify(obj: any, options?: JSStringifyOptions): string {
  const {
    space = 0,
    propertyReplacer = null,
    valueReplacer = {},
    quotes = 'single',
    inlineArrayLimit = undefined,
    camelCase = false,
    camelCaseFn = camelCaseTransform,
    propertyRenameMap = {},
    defaultValuesMap = {},
    defaultValuesSetter = {},
    include = [],
    exclude = [],
    json = false  // JSON compliance mode
  } = options || {};

  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);

  let indentLevel: number = 0;

  const serialize = (root: any, obj: any, currentPath = ""): string => {
    if (valueReplacer[''] instanceof Function) {
      obj = valueReplacer['']({
        currentKey: '',
        currentPath,
        obj,
        propertyRenameMap,
        root,
        value: obj
      }); // Call valueReplacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map((item, index) => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(root, item, `${currentPath}/${index}`);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;

      // Apply custom default values setter if available
      Object.keys(valueReplacer).forEach(pattern => {
        if (matchesPattern(currentPath, dirname(pattern))) {
          const replacer: JSStringifyReplacer = valueReplacer[pattern];
          const property = pattern.split('/').pop(); // get the last segment as property name
          if (property === '*') {
            // Apply setter to all properties if needed
            Object.keys(obj).forEach(prop => {
              obj[prop] = replacer({
                currentKey: prop,
                currentPath: `${currentPath}/${prop}`,
                obj,
                value: obj[prop],
                root,
                propertyRenameMap
              });
            });
          } else if (property) {
            obj[property] = replacer({
              currentKey: property,
              currentPath,
              obj,
              value: obj[property],
              root,
              propertyRenameMap
            });
          }

        }
      });

      // Apply custom default values setter if available
      Object.keys(defaultValuesSetter).forEach(pattern => {
        if (matchesPattern(currentPath, dirname(pattern))) {
          const setter: JSStringifySetter = defaultValuesSetter[pattern];
          const property = pattern.split('/').pop(); // get the last segment as property name
          if (property === '*') {
            // Apply setter to all properties if needed
            Object.keys(obj).forEach(prop => {
              if (obj[prop] === undefined) {
                obj[prop] = setter({
                  currentKey: prop,
                  currentPath: `${currentPath}/${prop}`,
                  obj,
                  value: obj[prop],
                  root,
                  defaultValuesMap
                });
              }
            });
          } else if (property && obj[property] === undefined) {
            obj[property] = setter({
              currentKey: property,
              currentPath,
              obj,
              value: obj[property],
              root,
              defaultValuesMap
            });
          }
        }
      });

      // Apply default values based on currentPath
      Object.keys(defaultValuesMap).forEach(pattern => {
        if (matchesPattern(currentPath, pattern)) {
          const property = pattern.split('/').pop(); // get the last segment as property name
          if (property === '*') {
            // Apply to all properties if needed
            Object.keys(obj).forEach(prop => {
              if (obj[prop] === undefined) {
                obj[prop] = defaultValuesMap[pattern];
              }
            });
          } else if (property && obj[property] === undefined) {
            obj[property] = defaultValuesMap[pattern];
          }
        }
      });


      // INCLUDES EXCLUDES

      obj = Object.keys(obj).reduce((m, key) => {
        const fullPath = `${currentPath}/${key}`;
        if (shouldInclude(fullPath, {
          exclude,
          include
        })) {
          m[key] = obj[key];
        }
        return m;
      }, {});


      // CHANGE PROPERTY NAMES

      const props = Object.keys(obj).map(key => {
        const fullPath = `${currentPath}/${key}`;
        const value = valueReplacer instanceof Function ? valueReplacer({
          currentKey: key,
          currentPath: fullPath,
          obj: obj[key],
          propertyRenameMap,
          root
        }) : obj[key];

        let replacedKey: string;
        switch (true) {
          case propertyReplacer instanceof Function:
            replacedKey = propertyReplacer({ root, obj, currentKey: key, currentPath: fullPath, propertyRenameMap });
            break;
          default:
            replacedKey = defaultPropertyReplacer(key, fullPath, propertyRenameMap);
        }

        const finalKey = camelCase ? camelCaseFn(replacedKey) : replacedKey;
        const keyPart = json ? `"${finalKey}"` : isSimpleKey(finalKey) ? finalKey : `"${finalKey}"`;
        const valuePart = serialize(root, value, fullPath);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return json ? `"${escapeStringForDoubleQuotes(obj)}"` : chooseQuotes(obj, quotes);
    } else {
      return String(obj);
    }
  };

  return serialize(obj, obj);
}
