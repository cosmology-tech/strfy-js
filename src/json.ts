import { camelCaseTransform, isSimpleKey } from "./utils";

export interface JSONStringifyOptions {
  space?: number; // defaults to 0
  replacer?: (this: any, key: string, value: any) => any; // defaults to null
  quotes?: 'single' | 'double'; // defaults to double
  inlineArrayLimit?: number; // default is undefined, allowing inline arrays below this length
  camelCase?: boolean; // defaults to false
  camelCaseFn?: (str: string) => string; // optional function to convert keys to camelCase
}

function escapeString(str: string): string {
  str = str.replace(/\\/g, '\\\\');  // Escape backslashes first
  str = str.replace(/\n/g, '\\n');  // Escape newlines
  str = str.replace(/\r/g, '\\r');  // Escape carriage returns
  str = str.replace(/\t/g, '\\t');  // Escape tabs
  str = str.replace(/"/g, '\\"');  // Escape double quotes
  return str;
}

function chooseQuotes(str: string): string {
  return `"${escapeString(str)}"`;
}

export function jsonStringify(obj: any, options?: JSONStringifyOptions): string {
  const { space = 0, replacer = null, quotes = 'double', inlineArrayLimit = undefined, camelCase = false, camelCaseFn = camelCaseTransform } = options || {};
  const isObject = (val: any): boolean => typeof val === 'object' && val !== null && !Array.isArray(val);
  let indentLevel: number = 0;

  const serialize = (obj: any): string => {
    if (replacer instanceof Function) {
      obj = replacer.call(null, '', obj); // Call replacer for the root element initially
    }

    if (Array.isArray(obj)) {
      const useInline = inlineArrayLimit !== undefined && obj.length <= inlineArrayLimit;
      indentLevel++;
      const result = '[' + (space && !useInline ? '\n' : '') + obj.map(item => {
        return ' '.repeat(useInline ? 0 : indentLevel * space) + serialize(item);
      }).join(',' + (space && !useInline ? '\n' : ' ')) + (space && !useInline ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + ']';
      indentLevel--;
      return result;
    } else if (isObject(obj)) {
      indentLevel++;
      const props = Object.keys(obj).map(key => {
        const value = replacer instanceof Function ? replacer.call(obj, key, obj[key]) : obj[key];
        const transformedKey = camelCase ? camelCaseFn(key) : key;
        const keyPart = isSimpleKey(transformedKey) ? transformedKey : chooseQuotes(transformedKey);
        const valuePart = serialize(value);
        return ' '.repeat(indentLevel * space) + `${keyPart}: ${valuePart}`;
      });
      const result = '{' + (space ? '\n' : '') + props.join(',' + (space ? '\n' : ' ')) + (space ? '\n' + ' '.repeat((indentLevel - 1) * space) : '') + '}';
      indentLevel--;
      return result;
    } else if (typeof obj === 'string') {
      return chooseQuotes(obj);
    } else {
      return String(obj);
    }
  };

  return serialize(obj);
}
