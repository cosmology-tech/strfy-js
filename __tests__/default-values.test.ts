import { jsStringify, JSStringifyOptions, JSStringifySetterOptions } from "../src";
import { Asset, AssetList } from '../test-utils';
import assetList from '../__fixtures__/assets.json';
import assetList2 from '../__fixtures__/assets2.json';

it('defaultValue', () => {
  const options: JSStringifyOptions = {
    camelCase: true,
    space: 2,
    defaultValuesSetter: {
      "/assets/*/type_asset": function (options: JSStringifySetterOptions<Asset, AssetList>): any {
        const asset = options.obj;
        switch (true) {
          case asset.base.startsWith('factory/'):
            return 'sdk.Factory';

          case asset.base.startsWith('ft') && options.root.chain_name === 'bitsong':
            return 'bitsong';

          case asset.base.startsWith('erc20/'):
            return 'erc.Token';

          case asset.base.startsWith('ibc/'):
            return 'ibc'

          case asset.base.startsWith('cw20:'):
            return 'cw20'

          default:
            return 'unknown'
        }
      }
    },
    propertyRenameMap: {
      '/assets/*/type_asset': 'asset_type'
    }
  };
  const jsonString = jsStringify(assetList, options);
  expect(jsonString).toMatchSnapshot();
});

it('defaultValue undefined', () => {
  const options: JSStringifyOptions = {
    camelCase: true,
    space: 2,
    defaultValuesSetter: {
      "/assets/*/type_asset": function (options: JSStringifySetterOptions<Asset, AssetList>): any {
        const asset = options.obj;
        switch (true) {
          case asset.base.startsWith('factory/'):
            return 'sdk.Factory';

          case asset.base.startsWith('ft') && options.root.chain_name === 'bitsong':
            return 'bitsong';

          case asset.base.startsWith('erc20/'):
            return 'erc.Token';

          case asset.base.startsWith('ibc/'):
            return 'ibc'

          case asset.base.startsWith('cw20:'):
            return 'cw20'

          default:
            return;
        }
      }
    },
    propertyRenameMap: {
      '/assets/*/type_asset': 'asset_type'
    }
  };
  const jsonString = jsStringify(assetList2, options);
  expect(jsonString).toMatchSnapshot();
});