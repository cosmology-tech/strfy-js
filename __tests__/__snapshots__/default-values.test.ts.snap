// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defaultValue 1`] = `
"{
  $schema: '../assetlist.schema.json',
  chainName: 'comdex',
  assets: [
    {
      description: 'Native Token of Comdex Protocol',
      denomUnits: [
        {
          denom: 'ucmdx',
          exponent: 0
        },
        {
          denom: 'cmdx',
          exponent: 6
        }
      ],
      base: 'ucmdx',
      name: 'Comdex',
      display: 'cmdx',
      symbol: 'CMDX',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg'
      },
      coingeckoId: 'comdex',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmdx.svg'
        }
      ],
      assetType: 'unknown'
    },
    {
      assetType: 'sdk.Factory',
      description: 'Governance Token of Harbor protocol on Comdex network',
      denomUnits: [
        {
          denom: 'uharbor',
          exponent: 0
        },
        {
          denom: 'harbor',
          exponent: 6
        }
      ],
      base: 'uharbor',
      name: 'Harbor',
      display: 'harbor',
      symbol: 'HARBOR',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg'
      },
      coingeckoId: 'harbor-2',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/harbor.svg'
        }
      ]
    },
    {
      description: 'Stable Token of Harbor protocol on Comdex network',
      denomUnits: [
        {
          denom: 'ucmst',
          exponent: 0
        },
        {
          denom: 'cmst',
          exponent: 6
        }
      ],
      base: 'ucmst',
      name: 'CMST',
      display: 'cmst',
      symbol: 'CMST',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
      },
      coingeckoId: 'composite',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
        }
      ],
      assetType: 'unknown'
    },
    {
      description: 'My ibc token',
      denomUnits: [
        {
          denom: 'uibc',
          exponent: 0
        },
        {
          denom: 'ibc',
          exponent: 6
        }
      ],
      base: 'ibc/A358FDF578F3483381C3191F86E4C6859E76100',
      name: 'IBC',
      display: 'ibc',
      symbol: 'IBC',
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
      },
      coingeckoId: 'ibc',
      images: [
        {
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.png',
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/comdex/images/cmst.svg'
        }
      ],
      assetType: 'ibc'
    }
  ]
}"
`;

exports[`defaultValue undefined 1`] = `
"{
  $schema: '../assetlist.schema.json',
  chainName: 'picasso',
  assets: [
    {
      description: 'The native staking and governance token of Picasso Kusama.',
      denomUnits: [
        {
          denom: 'ppica',
          exponent: 0
        },
        {
          denom: 'pica',
          exponent: 12
        }
      ],
      assetType: 'substrate',
      base: 'ppica',
      name: 'Pica',
      display: 'pica',
      symbol: 'PICA',
      logoURIs: {
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/composable/images/pica.svg'
      },
      images: [
        {
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/composable/images/pica.svg'
        }
      ]
    },
    {
      denomUnits: [
        {
          denom: '4',
          exponent: 0
        },
        {
          denom: 'ksm',
          exponent: 12
        }
      ],
      assetType: 'substrate',
      base: '4',
      name: 'Kusama',
      display: 'ksm',
      symbol: 'KSM',
      traces: [
        {
          type: 'bridge',
          counterparty: {
            chainName: 'kusama',
            baseDenom: 'Planck'
          },
          provider: 'Kusama Parachain'
        }
      ],
      images: [
        {
          imageSync: {
            chainName: 'kusama',
            baseDenom: 'Planck'
          },
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/kusama/images/ksm.svg'
        }
      ],
      logoURIs: {
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/kusama/images/ksm.svg'
      }
    },
    {
      denomUnits: [
        {
          denom: '79228162514264337593543950342',
          exponent: 0
        },
        {
          denom: 'dot',
          exponent: 10
        }
      ],
      assetType: 'substrate',
      base: '79228162514264337593543950342',
      name: 'Polkadot',
      display: 'dot',
      symbol: 'DOT',
      traces: [
        {
          type: 'ibc',
          counterparty: {
            chainName: 'composablepolkadot',
            baseDenom: '79228162514264337593543950342',
            channelId: 'channel-15'
          },
          chain: {
            channelId: 'channel-15',
            path: 'transfer/channel-15/79228162514264337593543950342'
          }
        }
      ],
      images: [
        {
          imageSync: {
            chainName: 'composablepolkadot',
            baseDenom: '79228162514264337593543950342'
          },
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/polkadot/images/dot.svg',
          png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/polkadot/images/dot.png'
        }
      ],
      logoURIs: {
        png: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/polkadot/images/dot.png',
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/polkadot/images/dot.svg'
      }
    },
    {
      denomUnits: [
        {
          denom: '130',
          exponent: 0
        },
        {
          denom: 'usdt',
          exponent: 6
        }
      ],
      assetType: 'substrate',
      base: '130',
      name: 'Statemine',
      display: 'usdt',
      symbol: 'USDT',
      traces: [
        {
          type: 'bridge',
          counterparty: {
            chainName: 'statemine',
            baseDenom: '130'
          },
          provider: 'Kusama Parachain'
        }
      ],
      images: [
        {
          imageSync: {
            chainName: 'statemine',
            baseDenom: '130'
          }
        }
      ]
    },
    {
      denomUnits: [
        {
          denom: '2125',
          exponent: 0
        },
        {
          denom: 'TNKR',
          exponent: 12
        }
      ],
      base: '2125',
      name: 'Tinkernet',
      display: 'TNKR',
      symbol: 'TNKR',
      traces: [
        {
          type: 'bridge',
          counterparty: {
            chainName: 'tinkernet',
            baseDenom: 'Planck'
          },
          provider: 'Tinkernet Parachain'
        }
      ],
      images: [
        {
          imageSync: {
            chainName: 'tinkernet',
            baseDenom: 'Planck'
          },
          svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/tinkernet/images/tnkr.svg'
        }
      ],
      logoURIs: {
        svg: 'https://raw.githubusercontent.com/cosmos/chain-registry/master/_non-cosmos/tinkernet/images/tnkr.svg'
      }
    }
  ]
}"
`;
